File: frontend/mobile/lib/services/time.service.dart
import 'package:dio/dio.dart';

import '../interfaces/time.interface.dart';
import 'api.service.dart';

class TimeException implements Exception {
  final String message;
  final String? code;
  final StudySession? session;

  const TimeException(this.message, {this.code, this.session});

  @override
  String toString() => message;
}

abstract class TimeRepository {
  Future<TimerSetup> getSetup();
  Future<ActiveStudySession> getActiveSession();
  Future<StudyDashboard> getDashboard();
  Future<StudySession> startSession({
    required int scheduleTimeId,
    required int studyTypeId,
  });
  Future<StudySession> pauseSession(int studyTimeId, int version);
  Future<StudySession> resumeSession(int studyTimeId, int version);
  Future<StudySession> finishSession(int studyTimeId, int version);
  Future<StudySession> heartbeatSession(int studyTimeId, int version);
  Future<StudySession> recoverSession(
    int studyTimeId,
    int version,
    String action,
  );
}

class TimeService implements TimeRepository {
  final ApiService _apiService;

  TimeService({ApiService? apiService})
    : _apiService = apiService ?? ApiService();

  static const _errorMessages = <String, String>{
    'NO_CURRENT_TERM': 'ยังไม่มีเทอมปัจจุบัน กรุณาสร้างเทอมก่อนเริ่มจับเวลา',
    'INVALID_TIMER_SELECTION': 'กรุณาเลือกวิชาและวิธีทบทวนให้ครบถ้วน',
    'SUBJECT_NOT_FOUND': 'ไม่พบวิชานี้ในตารางเรียนของเทอมปัจจุบัน',
    'STUDY_TYPE_NOT_FOUND': 'ไม่พบวิธีทบทวนที่เลือก',
    'OPEN_SESSION_EXISTS': 'มีรายการจับเวลาที่ยังจัดการไม่เสร็จอยู่แล้ว',
    'SESSION_NOT_FOUND': 'ไม่พบรายการจับเวลานี้ในเทอมปัจจุบัน',
    'SESSION_VERSION_CONFLICT':
        'รายการนี้ถูกแก้ไขจากอุปกรณ์อื่น ระบบจะโหลดข้อมูลล่าสุดให้',
    'INVALID_SESSION_STATE': 'สถานะรายการจับเวลาไม่รองรับคำสั่งนี้',
    'SESSION_HARD_LIMIT_REACHED':
        'รายการนี้ถึงเวลาสูงสุด 4 ชั่วโมงแล้ว กรุณาบันทึกหรือยกเลิก',
    'SESSION_NOT_STALE': 'รายการนี้ไม่จำเป็นต้องกู้คืนแล้ว',
    'INVALID_SESSION_REQUEST': 'ข้อมูลรายการจับเวลาไม่ถูกต้อง',
    'INVALID_RECOVERY_REQUEST': 'ตัวเลือกการกู้รายการจับเวลาไม่ถูกต้อง',
  };

  @override
  Future<TimerSetup> getSetup() async {
    try {
      final response = await _apiService.get('/user/time/setup');
      return TimerSetup.fromJson(_body(response.data));
    } on DioException catch (error) {
      throw _toException(error, 'ไม่สามารถโหลดข้อมูลสำหรับจับเวลาได้');
    } on TypeError {
      throw const TimeException('ข้อมูลสำหรับจับเวลาจากระบบไม่ถูกต้อง');
    }
  }

  @override
  Future<ActiveStudySession> getActiveSession() async {
    try {
      final response = await _apiService.get('/user/time/active');
      return ActiveStudySession.fromJson(_body(response.data));
    } on DioException catch (error) {
      throw _toException(error, 'ไม่สามารถตรวจสอบรายการที่กำลังจับเวลาได้');
    } on TypeError {
      throw const TimeException('ข้อมูลรายการจับเวลาไม่ถูกต้อง');
    }
  }

  @override
  Future<StudyDashboard> getDashboard() async {
    try {
      final response = await _apiService.get('/user/time/dashboard');
      return StudyDashboard.fromJson(_body(response.data));
    } on DioException catch (error) {
      throw _toException(error, 'ไม่สามารถโหลดสถิติการทบทวนได้');
    } on TypeError {
      throw const TimeException('ข้อมูลสถิติการทบทวนไม่ถูกต้อง');
    }
  }

  @override
  Future<StudySession> startSession({
    required int scheduleTimeId,
    required int studyTypeId,
  }) async {
    try {
      final response = await _apiService.post(
        '/user/time/start',
        data: {
          'schedule_time_id': scheduleTimeId,
          'study_type_id': studyTypeId,
        },
      );
      return _session(response.data);
    } on DioException catch (error) {
      throw _toException(error, 'ไม่สามารถเริ่มจับเวลาได้');
    }
  }

  @override
  Future<StudySession> pauseSession(int studyTimeId, int version) =>
      _patchSession(studyTimeId, 'pause', {
        'version': version,
      }, 'ไม่สามารถพักเวลาได้');

  @override
  Future<StudySession> resumeSession(int studyTimeId, int version) =>
      _patchSession(studyTimeId, 'resume', {
        'version': version,
      }, 'ไม่สามารถจับเวลาต่อได้');

  @override
  Future<StudySession> finishSession(int studyTimeId, int version) =>
      _patchSession(studyTimeId, 'finish', {
        'version': version,
      }, 'ไม่สามารถบันทึกเวลาทบทวนได้');

  @override
  Future<StudySession> heartbeatSession(int studyTimeId, int version) async {
    try {
      final response = await _apiService.post(
        '/user/time/$studyTimeId/heartbeat',
        data: {'version': version},
      );
      return _session(response.data);
    } on DioException catch (error) {
      throw _toException(error, 'ไม่สามารถซิงก์รายการจับเวลาได้');
    }
  }

  @override
  Future<StudySession> recoverSession(
    int studyTimeId,
    int version,
    String action,
  ) => _patchSession(studyTimeId, 'recover', {
    'version': version,
    'action': action,
  }, 'ไม่สามารถจัดการรายการจับเวลาที่ค้างอยู่ได้');

  Future<StudySession> _patchSession(
    int studyTimeId,
    String action,
    Map<String, dynamic> data,
    String fallback,
  ) async {
    try {
      final response = await _apiService.dio.patch(
        '/user/time/$studyTimeId/$action',
        data: data,
      );
      return _session(response.data);
    } on DioException catch (error) {
      throw _toException(error, fallback);
    }
  }

  StudySession _session(dynamic responseData) {
    final data = _body(responseData)['data'];
    if (data is! Map) {
      throw const TimeException('ข้อมูลรายการจับเวลาไม่ถูกต้อง');
    }
    return StudySession.fromJson(Map<String, dynamic>.from(data));
  }

  Map<String, dynamic> _body(dynamic value) =>
      Map<String, dynamic>.from(value as Map);

  TimeException _toException(DioException error, String fallback) {
    final data = error.response?.data;
    if (data is Map) {
      final body = Map<String, dynamic>.from(data);
      final code = body['code']?.toString();
      final rawSession = body['data'];
      return TimeException(
        code == null
            ? body['message']?.toString() ?? fallback
            : _errorMessages[code] ?? body['message']?.toString() ?? fallback,
        code: code,
        session: rawSession is Map
            ? StudySession.fromJson(Map<String, dynamic>.from(rawSession))
            : null,
      );
    }
    return TimeException(fallback);
  }
}

File: frontend/web/user/src/services/time.service.ts
import axios, { AxiosInstance } from "axios";
import Cookies from "js-cookie";
import type {
  ActiveSessionResponse,
  RecoverStudySessionRequest,
  SessionResponse,
  SessionVersionRequest,
  StartStudySessionRequest,
  StudyDashboard,
  StudySession,
  TimeApiErrorPayload,
  TimerSetupResponse,
} from "@/interfaces/time.interface";

const thaiErrorMessages: Record<string, string> = {
  NO_CURRENT_TERM: "ยังไม่มีเทอมปัจจุบัน กรุณาสร้างเทอมก่อนเริ่มจับเวลา",
  INVALID_TIMER_SELECTION: "กรุณาเลือกวิชาและวิธีทบทวนให้ครบถ้วน",
  SUBJECT_NOT_FOUND: "ไม่พบวิชานี้ในตารางเรียนของเทอมปัจจุบัน",
  STUDY_TYPE_NOT_FOUND: "ไม่พบวิธีทบทวนที่เลือก",
  OPEN_SESSION_EXISTS: "มีรายการจับเวลาที่ยังจัดการไม่เสร็จอยู่แล้ว",
  SESSION_NOT_FOUND: "ไม่พบรายการจับเวลานี้ในเทอมปัจจุบัน",
  SESSION_VERSION_CONFLICT:
    "รายการนี้ถูกแก้ไขจากหน้าต่างหรืออุปกรณ์อื่น ระบบโหลดข้อมูลล่าสุดให้แล้ว",
  INVALID_SESSION_STATE: "สถานะรายการจับเวลาไม่รองรับคำสั่งนี้",
  SESSION_HARD_LIMIT_REACHED:
    "รายการนี้ถึงเวลาสูงสุด 4 ชั่วโมงแล้ว กรุณายืนยันว่าจะบันทึกหรือยกเลิก",
  SESSION_NOT_STALE: "รายการนี้ไม่จำเป็นต้องกู้คืนแล้ว",
  INVALID_SESSION_REQUEST: "ข้อมูลรายการจับเวลาไม่ถูกต้อง กรุณาลองใหม่",
  INVALID_RECOVERY_REQUEST: "ตัวเลือกการกู้รายการจับเวลาไม่ถูกต้อง",
};

export class TimeApiError extends Error {
  status?: number;
  code?: string;
  session?: StudySession;

  constructor(
    message: string,
    options?: { status?: number; code?: string; session?: StudySession }
  ) {
    super(message);
    this.name = "TimeApiError";
    this.status = options?.status;
    this.code = options?.code;
    this.session = options?.session;
  }
}

class TimeService {
  private apiClient: AxiosInstance;

  constructor() {
    this.apiClient = axios.create({
      baseURL: process.env.NEXT_PUBLIC_API_URL || "http://localhost:4000",
      timeout: 10000,
    });

    this.apiClient.interceptors.request.use((config) => {
      const accessToken = Cookies.get("accessToken");
      if (accessToken) {
        config.headers.Authorization = `Bearer ${accessToken}`;
      }
      return config;
    });
  }

  async getSetup(): Promise<TimerSetupResponse> {
    try {
      const response = await this.apiClient.get<TimerSetupResponse>(
        "/user/time/setup"
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "ไม่สามารถโหลดข้อมูลสำหรับจับเวลาได้");
    }
  }

  async getActiveSession(): Promise<ActiveSessionResponse> {
    try {
      const response = await this.apiClient.get<ActiveSessionResponse>(
        "/user/time/active"
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "ไม่สามารถตรวจสอบรายการที่กำลังจับเวลาได้");
    }
  }

  async getDashboard(): Promise<StudyDashboard> {
    try {
      const response = await this.apiClient.get<StudyDashboard>(
        "/user/time/dashboard"
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "ไม่สามารถโหลดสถิติการทบทวนได้");
    }
  }

  async startSession(data: StartStudySessionRequest): Promise<SessionResponse> {
    try {
      const response = await this.apiClient.post<SessionResponse>(
        "/user/time/start",
        data
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "ไม่สามารถเริ่มจับเวลาได้");
    }
  }

  async pauseSession(
    studyTimeId: number,
    data: SessionVersionRequest
  ): Promise<SessionResponse> {
    return this.patchSession(studyTimeId, "pause", data, "ไม่สามารถพักเวลาได้");
  }

  async resumeSession(
    studyTimeId: number,
    data: SessionVersionRequest
  ): Promise<SessionResponse> {
    return this.patchSession(studyTimeId, "resume", data, "ไม่สามารถจับเวลาต่อได้");
  }

  async finishSession(
    studyTimeId: number,
    data: SessionVersionRequest
  ): Promise<SessionResponse> {
    return this.patchSession(
      studyTimeId,
      "finish",
      data,
      "ไม่สามารถบันทึกเวลาทบทวนได้"
    );
  }

  async heartbeatSession(
    studyTimeId: number,
    data: SessionVersionRequest
  ): Promise<SessionResponse> {
    try {
      const response = await this.apiClient.post<SessionResponse>(
        `/user/time/${studyTimeId}/heartbeat`,
        data
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, "ไม่สามารถซิงก์รายการจับเวลาได้");
    }
  }

  async recoverSession(
    studyTimeId: number,
    data: RecoverStudySessionRequest
  ): Promise<SessionResponse> {
    return this.patchSession(
      studyTimeId,
      "recover",
      data,
      "ไม่สามารถจัดการรายการจับเวลาที่ค้างอยู่ได้"
    );
  }

  private async patchSession(
    studyTimeId: number,
    action: string,
    data: SessionVersionRequest | RecoverStudySessionRequest,
    fallbackMessage: string
  ): Promise<SessionResponse> {
    try {
      const response = await this.apiClient.patch<SessionResponse>(
        `/user/time/${studyTimeId}/${action}`,
        data
      );
      return response.data;
    } catch (error) {
      throw this.toError(error, fallbackMessage);
    }
  }

  private toError(error: unknown, fallbackMessage: string): TimeApiError {
    if (axios.isAxiosError<TimeApiErrorPayload>(error)) {
      const payload = error.response?.data;
      const code = payload?.code;
      const message =
        (code ? thaiErrorMessages[code] : undefined) ||
        payload?.message ||
        (error.code === "ERR_NETWORK"
          ? "ไม่สามารถเชื่อมต่อเซิร์ฟเวอร์ได้ กรุณาตรวจสอบอินเทอร์เน็ต"
          : error.message) ||
        fallbackMessage;
      return new TimeApiError(message, {
        status: error.response?.status,
        code,
        session: payload?.data,
      });
    }
    return new TimeApiError(
      error instanceof Error ? error.message : fallbackMessage
    );
  }
}

export const timeService = new TimeService();
export default timeService;
